src/gammalasso.jl

```diff
@@ -0,0 +1,406 @@
+end
+
+# ## Error handling, holds a copy of the current path to debug
+# immutable ConvergenceException <: Exception
+#     msg
+#     debugvars
+# end
+#
+
+## MODEL CONSTRUCTION
+
+"Compute coeffiecient specific weights vector ω_j^t based on previous iteration coefficients β"
+function computeω!{T}(ω::Vector{T}, γ::Vector{T}, β::SparseCoefficients{T})
+    # initialize to a vector of ones
+    fill!(ω,one(T))
+
+    # set weights of non zero betas
+    @inbounds @simd for icoef = 1:nnz(β)
+        ipred = β.coef2predictor[icoef]
+        γi = γ[ipred]
+        if γi != 0.0
+            ω[ipred] = 1.0/(1.0+γi*abs(β.coef[icoef]))
+        end
+    end
+
+    # rescale(ω,p) # not sure if rescaling is the right thing to do here, nothing about it in Taddy (2016)
+    nothing
+end
+
+function StatsBase.fit{T<:AbstractFloat,V<:FPVector}(::Type{GammaLassoPath},
+                                                     X::AbstractMatrix{T}, y::V, d::UnivariateDistribution=Normal(),
+                                                     l::Link=canonicallink(d);
+                                                     γ::@compat(Union{Number,Vector{Number}})=0.0,
+                                                     wts::@compat(Union{FPVector,Void})=ones(T, length(y)),
+                                                     offset::AbstractVector=similar(y, 0),
+                                                     α::Number=one(eltype(y)), nλ::Int=100,
+                                                     λminratio::Number=ifelse(size(X, 1) < size(X, 2), 0.01, 1e-4),
+                                                     λ::@compat(Union{Vector,Void})=nothing, standardize::Bool=true,
+                                                     intercept::Bool=true,
+                                                     naivealgorithm::Bool=(!isa(d, Normal) || !isa(l, IdentityLink) || size(X, 2) > 5*size(X, 1)),
+                                                     dofit::Bool=true,
+                                                     irls_tol::Real=1e-7, randomize::Bool=RANDOMIZE_DEFAULT,
+                                                     maxncoef::Int=min(size(X, 2), 2*size(X, 1)), fitargs...)
+
+    size(X, 1) == size(y, 1) || DimensionMismatch("number of rows in X and y must match")
+    n = length(y)
+    length(wts) == n || error("length(wts) = $(length(wts)) should be 0 or $n")
+
+    # Standardize predictors if requested
+    if standardize
+        Xnorm = vec(full(std(X, 1, corrected=false)))
+        for i = 1:length(Xnorm)
+            @inbounds Xnorm[i] = 1/Xnorm[i]
+        end
+        X = scale(X, Xnorm)
+    else
+        Xnorm = T[]
+    end
+
+    # gamma lasso adaptation
+    # can potentially pass a different γ for each element of X, but if scalar we copy it to all params
+    p = size(X, 2)
+    if length(γ) == 1
+      γ=convert(Vector{T},repmat([γ],p))
+    else
+      @assert length(γ)==p "length(γ) != number of parameters ($p)"
```

Comment:
`@assert` is generally supposed to be used to make sure the programmer didn't make a mistake, not the user. Better to write this as `length(y) == p || throw(DimensionMismatch("length(γ) != number of parameters ($p)")))`